pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,332 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package operatingsystemconfig
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pelletier/go-toml"
+	"k8s.io/utils/pointer"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsv1alpha1helper "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1/helper"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	"github.com/gardener/gardener/pkg/utils/retry"
+	"github.com/gardener/gardener/pkg/utils/structuredmap"
+)
+
+// ReconcileContainerdConfig sets required values of the given containerd configuration.
+func (r *Reconciler) ReconcileContainerdConfig(ctx context.Context, log logr.Logger, criConfig *extensionsv1alpha1.CRIConfig, containerdChanges containerd) error {
+	if !extensionsv1alpha1helper.HasContainerdConfiguration(criConfig) {
+		return nil
+	}
+
+	if err := r.ensureContainerdConfigDirectories(); err != nil {
+		return err
+	}
```

Comment:
So far the `/etc/containerd/certs.d` dir was ensured by Gardener even when there is no containerd configuration: specified.https://github.com/gardener/gardener/blob/78146b200daa40fa0b77b5758b929d9d5bedaa91/docs/usage/containerd-registry-configuration.md#L63

With the above early exit it will be ensured only when containerd config is specified in the OSC. Is the containerd config always set?

Reply:
Gardener defaults the CRI to `containerd`, see https://github.com/gardener/gardener/blob/078a6b88d065567f6d1285f370932c20bb23f021/pkg/apis/core/v1beta1/defaults_shoot.go#L55 and it's the only valid value that can be set/defaulted today. So yes, it should always be set in the OSC.

Reply:
Okay, it is https://github.com/gardener/gardener/blob/078a6b88d065567f6d1285f370932c20bb23f021/pkg/apis/core/v1beta1/defaults_shoot.go#L55 + https://github.com/gardener/gardener/blob/832247d1a21dc3473f44ffc2206b2e8056a54384/pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go#L945-L949 that make sure that the criConfig will be set for an OSC.